// aframe-map/lib/entity.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = aframeMapComponent;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// Inspired by https://github.com/flysonic10/aframe-passthrough
var extendDeep = window.AFRAME.utils.extendDeep;
var meshMixin = window.AFRAME.primitives.getMeshMixin();

/**
 * @param aframe {Object} The Aframe instance to register with
 * @param componentName {String} The component name to use
 */
function aframeMapComponent(aframe, componentName) {
  var _defaultComponents;

  window.AFRAME.registerPrimitive('a-' + componentName, extendDeep({}, meshMixin, {
    defaultComponents: (_defaultComponents = {}, _defineProperty(_defaultComponents, componentName, {}), _defineProperty(_defaultComponents, 'geometry', {
      primitive: 'plane'
    }), _defineProperty(_defaultComponents, 'material', {
      color: '#ffffff',
      shader: 'flat',
      side: 'both',
      transparent: true
    }), _defaultComponents),

    mappings: {
      height: 'geometry.height',
      width: 'geometry.width'
    }
  }));
}
